text/0083-component.md

```diff
@@ -0,0 +1,118 @@
+# Component
+
+Introducing the notion of Component as a first class concept in OpenTelemetry
+which describes the named `Tracer|Meter` in the data model.
+
+## Motivation
+
+The main motivation for this is to better model telemetry coming from different
+instrumented or instrumenting libraries inside an Application (`Resource`). This
+change affects only the OpenTelemetry protocol and exporters, and it does not
+require any API changes.
+
+Currently OpenTelemetry has defined a semantic convention for the `component`
+Span attribute which already caused some concerns and questions, see [here](
+https://github.com/open-telemetry/opentelemetry-specification/issues/336)
+
+The proposal is to make `Component` a first class concept in OpenTelemetry, with
+a scope defined by the `named Tracer|Meter`. It describes all telemetry
+generated by one `named Tracer|Meter`.
+
+## Explanation
+
+```
+                         Application
++-----------------------------------------------------------------+
+|                   TracerProvider(Resource)                      |
+|                   MeterProvider(Resource)                       |
+|                                                                 |
+|      Instrumented Library 1         Instrumented Library 2      |
```

Comment:
Your diagram is wrong. While I agree that for "components" it would be ideal to break up the application by (instrumented) library, named tracers, by design, break it up by instrumentation library. I fear we cannot really use named tracers for what you suggest.

Reply:
What is the mapping between instrumentation library and instrumented library? I thought that it is 1:1 (I do agree that naming can be different).

@dyladan to your point in case of a JDBC instrumentation the instrumented library is JDBC and the fact that connects to a mysql or postgress is a property of the instance of the JDBC library which can be captured as suggested in the extra attributes that we can add later.

Reply:
It's an n:m mapping. Granted, most often it will be 1:1, but in the general case you need to consider:
* An instrumenting library might instrument multiple "components". E.g. an auto instrumentation may instrument all implementations of a DB driver interface, so it covers mysql, postgres, oracle, ...
* There may be multiple instrumentations for the same library. E.g.
  * an opensource library instrumenting some component gets abandoned, then forked under a new name. 
  * a library to instrument some other library is created internally, at different companies under different names. Then one of them or more are published/opensourced under different names or an opensource/commercial implementation is developed independently.
  * a library might start out being instrumented by a 3rd party vendor, but then the instrumentation is built right into the library.
  * for some component, both an auto-instrumentation and wrapper solution might exist (since auto-instrumentation is not always possible or wanted, both have their advantages and disadvantages).

Depending on what is meant by the vague term "component" there may be an additional dimension that is relevant:
* A library may be used multiple times in the same application, resulting in multiple logical components, e.g. a MariaDB client library that is used to connect to maintain connections to both MariaDB and MySQL, maybe in the same process. You might want to assign each connection instance its own "component" with info like connection string, db name, ...

Named tracers are per instrumenting library, not per component or instrumented library.

Reply:
I would add another example for 1:n mapping. Consider a package released by a community member named `otel-db-all-in-one` which claims to instrument all major database libraries. It may auto-patch mysql, postgres, sqlite, mongo, etc. In this case, the Tracer Name would be `otel-db-all-in-one`, but the `component` should be the instrumented db lib, not the all in one instrumentation lib.

Reply:
Based on what I am reading, this is what I understand so far:
* It may be possible to have multiple instrumentation library instrumenting the same library (N : 1). This is why we include the package name (namespace) in the named tracer/meter name to distinguish between them.
* For the example where one released artifact/package instruments multiple libraries (like the dbs example) why do we not have one named tracer per db? It is probably important to have the ability to control the instrumentation at the db level not at the instrumenting package level. So I am not sure I understand that this is a valid use of the named Tracer.

Maybe we need to update the named Tracer/Meter to say that it applies to only one instrumented library/framework?

Reply:
@bogdandrutu 
ad 1: correct
ad 2: I agree that it might make sense to be able to control it at a DB level. This is, however, a different use case than the one addressed by named tracers. The use case covered by them is that a certain instrumentation version misbehaves or should be superseded and therefore disabled. The misbehavior/bug occurring in that instrumentation library version would be the same over all DBs. Same goes with superseding it; if I have a better instrumentation available I'd most probably want to have the old one replaced for all DBs.

Feel free to propose a means to define the instrument*ed* component in order to use it for further configuration but please accept that this is not what named tracers are (currently) about. The covered use case as mentioned above is clearly stated in the named tracer OTEP (https://github.com/open-telemetry/oteps/blob/master/text/0016-named-tracers.md) and the meaning of the provided instrumentation library name is clearly stated in the spec (https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#obtaining-a-tracer).

Reply:
I think this is now outdated.